Register Azure Service Bus health checks with factories

JoshKeegan opened the pull request:
As done for SQL Server in #175 

This allows for you to use `services.AddAzureServiceBusTopic(...`, `services.AddAzureServiceBusQueue(...` and `services.AddAzureEventHub(...` if you use dependency injection for your configuration. 

Sample usage:
```cs
services.AddAzureServiceBusQueue(
  provider => provider.GetService<SasmpleQueueConfig>().ConnStr, 
  provider => provider.GetService<SampleQueueConfig>().QueueName, 
  tags: new string[] { "liveness" });
```

Please let me know your thoughts on having two factories for these (one for the connection string, and one for the name). The alternative is for there to be a single factory that returns some new object with both the connection string and queue name on, but I wasn't sure what would be preferred.

sungam3r commented:
@JoshKeegan Please rebase onto `master` if you are going to continue work on this PR.

JoshKeegan commented:
Happy to rebase this @sungam3r but I don't think there was any more work to be done. Let me know if there is interest in merging it & I'll do the rebase.

sungam3r commented:
The fact is that I can not merge this PR while it has conflicts and did not pass tests. If you want this code to be merged, then I will review after rebasing.

JoshKeegan commented:
@sungam3r rebased on master. `AzureServiceBusHealthCheckBuilderExtensions` had quite a few changes and new extension methods added to it in master, so the merge was bad & I've just re-done the changes there from scratch, covering the new methods too.

The unit tests all merged cleanly so I've kept what I'd already written, but haven't yet added any new ones for the additional methods that have been added in master since the original PR was raised. Can you review the proposed changes now they'll merge cleanly and see if you're happy with them?

If you're happy with the idea then I can add the additional tests for the new methods now in master, but I'd like to know the proposed change is ok before investing further time in this.

sungam3r commented on src/HealthChecks.AzureServiceBus/DependencyInjection/AzureServiceBusHealthCheckBuilderExtensions.cs:
```diff
@@ -139,11 +213,60 @@ public static class AzureServiceBusHealthCheckBuilderExtensions
+        /// <param name="builder">The <see cref="IHealthChecksBuilder"/>.</param>
+        /// <param name="endpointFactory">A factory to build the azure service bus endpoint to be used, format sb://myservicebus.servicebus.windows.net/.</param>
+        /// <param name="queueNameFactory">A factory to build the name of the queue to check.</param>
+        /// <param name="tokenCredentialFactory">A factory to build the token credential for auth</param>
```
```suggestion
        /// <param name="tokenCredentialFactory">A factory to build the token credential for auth.</param>
```

JoshKeegan replied:
done

sungam3r commented on src/HealthChecks.AzureServiceBus/DependencyInjection/AzureServiceBusHealthCheckBuilderExtensions.cs:
```diff
@@ -187,14 +348,13 @@ public static class AzureServiceBusHealthCheckBuilderExtensions
         /// <param name="endpoint">The azure service bus endpoint to be used, format sb://myservicebus.servicebus.windows.net/.</param>
         /// <param name="topicName">The topic name of the topic to check.</param>
-        /// <param name="tokenCredential">The token credential for auth
+        /// <param name="tokenCredential">The token credential for auth</param>
```
```suggestion
        /// <param name="tokenCredential">The token credential for auth.</param>
```

sungam3r commented on src/HealthChecks.AzureServiceBus/DependencyInjection/AzureServiceBusHealthCheckBuilderExtensions.cs:
```diff
@@ -205,11 +365,60 @@ public static class AzureServiceBusHealthCheckBuilderExtensions
+        /// <param name="builder">The <see cref="IHealthChecksBuilder"/>.</param>
+        /// <param name="endpointFactory">A factory to build the azure service bus endpoint to be used, format sb://myservicebus.servicebus.windows.net/.</param>
+        /// <param name="topicNameFactory">A factory to build the topic name of the topic to check.</param>
+        /// <param name="tokenCredentialFactory">A factory to build the token credential for auth</param>
```
```suggestion
        /// <param name="tokenCredentialFactory">A factory to build the token credential for auth.</param>
```

sungam3r commented on src/HealthChecks.AzureServiceBus/DependencyInjection/AzureServiceBusHealthCheckBuilderExtensions.cs:
```diff
@@ -255,16 +512,14 @@ public static class AzureServiceBusHealthCheckBuilderExtensions
-        /// <param name="topicName">The topic name of the topic to check.</param>
         /// <param name="subscriptionName">The subscription name of the topic to check.</param>
-        /// <param name="tokenCredential">The token credential for auth
+        /// <param name="tokenCredential">The token credential for auth</param>
```
```suggestion
        /// <param name="tokenCredential">The token credential for auth.</param>
```

sungam3r commented on src/HealthChecks.AzureServiceBus/DependencyInjection/AzureServiceBusHealthCheckBuilderExtensions.cs:
```diff
@@ -276,11 +531,68 @@ public static class AzureServiceBusHealthCheckBuilderExtensions
+        /// <param name="endpointFactory">A factory to build the azure service bus endpoint to be used, format sb://myservicebus.servicebus.windows.net/.</param>
+        /// <param name="topicNameFactory">A factory to build the topic name of the topic to check.</param>
+        /// <param name="subscriptionNameFactory">A factory to build the subscription name of the topic to check.</param>
+        /// <param name="tokenCredentialFactory">A factory to build the token credential for auth</param>
```
```suggestion
        /// <param name="tokenCredentialFactory">A factory to build the token credential for auth.</param>
```

JoshKeegan replied:
done

sungam3r commented on test/HealthChecks.AzureServiceBus.Tests/DependencyInjection/AzureServiceBusQueueUnitWithTokenRegistrationTests.cs:
```diff
@@ -61,5 +61,34 @@ public void fail_when_no_health_check_configuration_provided()
+                        return "hubName";
+                    });
+
+            var serviceProvider = services.BuildServiceProvider();
```
```suggestion
            using var serviceProvider = services.BuildServiceProvider();
```

JoshKeegan commented on src/HealthChecks.AzureServiceBus/DependencyInjection/AzureServiceBusHealthCheckBuilderExtensions.cs:
```diff
@@ -139,11 +213,60 @@ public static class AzureServiceBusHealthCheckBuilderExtensions
             HealthStatus? failureStatus = default,
             IEnumerable<string>? tags = default,
+            TimeSpan? timeout = default) =>
+            builder.AddAzureServiceBusQueue(
```
done

JoshKeegan commented on test/HealthChecks.AzureServiceBus.Tests/DependencyInjection/AzureServiceBusQueueRegistrationTests.cs:
```diff
@@ -60,5 +59,34 @@ public void fail_when_no_health_check_configuration_provided()
+                        return "queueName";
+                    });
+
+            var serviceProvider = services.BuildServiceProvider();
```
done & applied to existing tests

JoshKeegan commented on test/HealthChecks.AzureServiceBus.Tests/DependencyInjection/AzureServiceBusTopicRegistrationTests.cs:
```diff
@@ -59,5 +59,34 @@ public void fail_when_no_health_check_configuration_provided()
+                        return "topicName";
+                    });
+
+            var serviceProvider = services.BuildServiceProvider();
```
done & applied to existing tests

JoshKeegan commented:
@sungam3r thanks for the review. Changes made, rebased on master & I've also added the additional unit tests I wanted.
So I'm now happy with this if you can re-review it.

sungam3r commented:
It looks like I lost thi PR from my open tabs, sorry. I'll review it tomorrow.
